File: T-Bank/Algorithms-and-Data-Structures-2025/1sprint-IntroductionToSorting-BinarySearch/contest/E-Корень_кубического_уравнения/solution.cpp
#include <iostream>
#include <cmath>
#include <iomanip>


double f(double& x, int& a, int& b, int&c, int& d) {
    return a  * std::pow(x, 3) + b * std::pow(x, 2) + c * x + d;
}


void solution() {
    /*
    Time Complexity O(log(10^8))
    Memory Complexity O(1)
    */
    int a, b, c, d;
    std::scanf("%d %d %d %d", &a, &b, &c, &d);
    double left = -100000000;
    double right = 100000000;
    double eplison = 0.00001;
    while (right - left > eplison) {
        double middle = left + (right - left) / 2.0;
        if (f(middle, a, b, c, d) == 0) {
            left = middle;
            right = middle;
            break;
        } else if (f(left, a, b, c, d) * f(middle, a, b, c, d) < 0) right = middle;
        else left = middle;
    }
    std::setprecision(5);
    std::cout << left + (right - left) / 2.0 << "\n";
}


int main() {
    solution();
}

File: T-Bank/Algorithms-and-Data-Structures-2025/1sprint-IntroductionToSorting-BinarySearch/contest/E-Корень_кубического_уравнения/solution.py
import sys


def f(x: float, a: int, b: int, c: int, d: int) -> float:
    return a * x**3.0 + b * x**2.0 + c * x + d


def solution() -> None:
    """
    Time Complexity O(log(10^9))
    Memory Complexity O(1)
    """
    a, b, c, d = map(int, sys.stdin.readline().rstrip().split())
    left: float = -10**10
    right: float = 10**10

    while right - left > 1e-8:
        middle: float = left + (right - left) / 2.0
        root: float = f(middle, a, b, c, d)
        if root == 0:
            left = right = middle
            break
        elif f(left, a, b, c, d) * f(middle, a, b, c, d) < 0: right = middle
        else: left = middle

    sys.stdout.write(f'{left + (right - left) / 2:.4f}\n')


if __name__ == '__main__':
    solution()